noir-projects/noir-contracts/contracts/app/closed_set_orderbook_contract/src/main.nr

```diff
@@ -0,0 +1,180 @@
+mod config;
+mod order_note;
+
+use aztec::macros::aztec;
+
+/// ## Overview
+/// This contract demonstrates how to implement an **Orderbook** that keeps all of the information in private. Orders
+/// are revealed only to a closed set of participants (those that know the Orderbook's secret keys).
+///
+/// **Note:**
+/// This is purely a demonstration implemented to test various features of Aztec.nr. The **Aztec team** does not
+/// consider this the optimal design for building a DEX.
+///
+/// ## Reentrancy Guard Considerations
+///
+/// ### 1. Private Functions:
+/// Reentrancy protection is typically necessary if entering an intermediate state that is only valid when
+/// the action completes uninterrupted. This follows the **Checks-Effects-Interactions** pattern.
+///
+/// - In this contract, **private functions** do not introduce intermediate states.
+///
+/// ### 2. Public Functions:
+/// The only public function in the contract is the constructor and there reentrancy is prevented by the
+/// `#[initializer]` macro.
+#[aztec]
+pub contract ClosedSetOrderbook {
+    use crate::{config::Config, order_note::OrderNote};
+    use aztec::{
+        macros::{functions::{initializer, private, public, utility}, storage::storage},
+        messages::logs::note::encode_and_encrypt_note,
+        note::{
+            lifecycle::destroy_note_unsafe, note_getter::get_note, retrieved_note::RetrievedNote,
+        },
+        oracle::random::random,
+        protocol_types::address::AztecAddress,
+        state_vars::{HasStorageSlot, Map, PrivateImmutable, SharedMutable},
+    };
+
+    use token::Token;
+
+    // Note that this value was copied over to e2e_closed_set_orderbook.test.ts. and needs to be kept in sync.
+    global CHANGE_CONFIG_DELAY: u64 = 60 * 60 * 24;
+
+    #[storage]
+    struct Storage<Context> {
+        // Using SharedMutable here reveals the whole config which is unfortunate.
+        // TODO: Implement smt. like `ClosedSharedMutable` as described in the comment below and use it here.
+        // https://github.com/AztecProtocol/aztec-packages/issues/14917#issuecomment-2960329494
+        config: SharedMutable<Config, CHANGE_CONFIG_DELAY, Context>,
+        orders: Map<Field, PrivateImmutable<OrderNote, Context>, Context>,
+    }
+
+    #[public]
+    #[initializer]
+    fn constructor(
+        token0: AztecAddress,
+        token1: AztecAddress,
+        fee_collector: AztecAddress,
+        fee: u16,
+    ) {
+        // TODO: Why don't we have a way to directly initialize SharedMutable when it's created? If this is done
+        // in the constructor then it's safe as we cannot interact with an undeployed contract that has initializer.
+        // Waiting for delay to pass here is quite annoying.
```

Comment:
Curious if reviewer thinks we would want to support something like this. When a contract needs to be initialized and the value is scheduled in `#[initializer]` then directly activating should be safe.